### PR TITLE
Support case-insensitive inline / scoped flags (?i) / (?i:…) in regex patterns [databricks][fast-ut][reduced-it]

### LINKED ISSUE #14947: [FEA] Support inline flag groups `(?i)` for case-insensitive matches in `regexp_like` / `regexp_extract` / `regexp_replace`
**Is your feature request related to a problem? Please describe.**
The RAPIDS Accelerator falls back to CPU for `RLIKE`, `REGEXP_EXTRACT`, and `REGEXP_REPLACE` whenever the pattern uses inline flag syntax such as `(?i)`. The RegexParser only special-cases `(?:...)` non-capturing groups; any other `(?` sequence that is not a lookahead is not supported as an inline flag group.

Three structurally distinct placements of `(?i)` appear in practice, each with different case-sensitivity scope:

  -- 1. Flag at the start -- entire pattern is case-insensitive
`col RLIKE '(?i)error message'` and `REGEXP_EXTRACT(col, '(?i)section:\s+(\w+)', 1)`

  -- 2. Flag mid-pattern -- only the suffix from the flag position onward is case-insensitive
`col RLIKE 'status":"(?i)ok'`

  -- 3. Flag inside an alternation group -- one alternative is case-insensitive, others are not
`col RLIKE 'type":"(active|inactive|(?i)pending)'`

**Describe the solution you'd like**
Extend the regex transpiler to handle inline flag groups for at least the (?i) (case-insensitive) flag, covering the three placement shapes above:

If only placement 1 is allowed, as it appears cudf already has some support for this, then it is a big win, but does not cover everything.

We might also be able to deal with some case-insensitive "constant" values by rewriting the regex, like `(?i)section:` could be rewritten to `[Ss][Ee][Cc][Tt][Ii][Oo][Nn]:`, `(?i)ok` to `[Oo][Kk]` and `(?i)pending` to `[Pp][Ee][Nn][Dd][Ii][Nn][Gg]`. But only if it is truly a constant value, or if we know that a character set (like \w) would not be impacted by `(?i)`.

### PR DESCRIPTION
**Pull request by igorpeshansky**
Fixes #14947.

### Description

`rlike` / `regexp_like`, `regexp_extract`, and `regexp_replace` fell back to the CPU whenever a pattern used inline or scoped flag syntax (`(?i)` or `(?i:…)`). The `RegexParser` only special-cased `(?:...)` non-capturing groups; any other `(?` sequence that was not a lookahead was rejected, so case-insensitive patterns never ran on the GPU.

This PR adds transpiler support for the case-insensitive flag, both the inline `(?i)` and the scoped (non-capturing) form `(?i:…)`. ASCII letters in case-insensitive runs (parts of the pattern where the `i` flag is in effect) are rewritten to match either case: literal characters and character-class ranges are folded (e.g. `(?i)abc` → `[aA][bB][cC]`, `(?i)[a-c]` → `[a-cA-C]`). Scoping follows Java's rules: an inline flag applies from its position to the end of the enclosing group, `(?-i)` turns it back off, and `(?i:...)` is only in scope within its group. This covers all three placement shapes from the issue: flag at the start of the pattern, mid-pattern, and inside an alternation branch. Results are unchanged (identical to the CPU); only the execution location changes (GPU instead of CPU fallback).

Anything that cannot be folded safely or easily still falls back to the CPU:

- Any positive flag other than `i` (e.g. `(?m)`, `(?s)`).
- A case-insensitive scope that would extend across alternatives, like `a(?i)b|c|d(?-i)e`, where carrying the flag state across `|` would require a significant rearchitecture of the transpiler. Note that the equivalent scoped form `a(?i:b)|(?i:c)|(?i:d)e` will run on the GPU.
- Case-insensitive matching of a letter-valued hex or octal escape such as `(?i)\x61`, which cannot be folded to a character class.

A negated non-`i` flag (e.g. `(?-m)`) is treated as a no-op, since those flags are off by default, and there is no mechanism to turn them on.

> [!TIP]
> The three implementation commits in this change (pre-refactoring, parser support, transpiler support) can be reviewed in order. Each builds independently.

#### Testing

- Parser unit tests for the bare and scoped inline-flag forms, including edge cases like `(?i-)`, `(?-)`, `(?-:)`, and `(?i-:)`.
- Transpiler unit tests for case-insensitive folding across find, replace, and split modes (literals, positive and negated character classes, nested and scoped flags), for `regexp_extract` group extraction, and for each fallback rule above.
- Rewrite-suite coverage confirming the fast path still applies through case-sensitive scoped-flag groups.
- Integration tests in `regexp_test.py` comparing GPU vs CPU for `rlike`, `regexp_extract`, and `regexp_replace` with case-insensitive inline and scoped flags, plus fallback coverage for unsupported flags, choice-crossing flags, and anchor combinations.

### Checklists

Documentation
- [x] Updated for new or modified user-facing features or behaviors
- [ ] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
Unrelated `RapidsStatisticsCollectionSuite` failure: `SPARK-33687: analyze all tables in a specific database`. Retrying.

build

**Comment by wjxiz1992** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala [hunk 1]
```diff
@@ -226,6 +229,163 @@ class RegularExpressionTranspilerSuite extends AnyFunSuite {
...
+    doTranspileTest("(?-s)abc", "abc")
+    doTranspileTest("(?i-s)abc", "[aA][bB][cC]")
+    // a trailing '-' with an empty negated set is accepted (Java allows `(?i-)`/`(?-)`)
+    doTranspileTest("(?i-)abc", "[aA][bB][cC]")
+    doTranspileTest("(?-)abc", "abc")
+  }
+
+  test("transpile case-insensitive inline flag in replace and split modes") {
+    // case folding is mode-independent, so (?i) must also work outside find mode
+    for (mode <- Seq(RegexReplaceMode, RegexSplitMode)) {
```
Comment (wjxiz1992) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:267]: [P2] Add end-to-end coverage for split consumers

This opts `RegexSplitMode` into the new flag folding, and `StringSplitRegExpMeta` feeds that mode to both `StringSplit` and `StringToMap`. The Python additions cover only `rlike`, `regexp_extract`, and `regexp_replace`, so these extra production paths are enabled without CPU/GPU parity coverage. Please add inline/scoped-flag cases to the existing split and `str_to_map` ITs (including limit and both delimiter variants), or keep `RegexSplitMode` on CPU in this PR.

File: integration_tests/src/main/python/regexp_test.py [hunk 2]
```diff
@@ -669,6 +669,18 @@ def test_regexp_extract():
                 'regexp_extract(a, "(a)(b)|(c)(d)", 4)'),
         conf=_regexp_conf)
 
+def test_regexp_extract_case_insensitive_inline_flag():
```
Comment (wjxiz1992) [inline integration_tests/src/main/python/regexp_test.py]: [P2] Extend the existing regex ITs

This uses the same unary DataFrame setup and `regexp_extract` expression shape as `test_regexp_extract` above. Per the repository test-coverage rule, please fold these queries into the existing test instead of adding another `def test_...`; the same applies to the new `rlike` and `regexp_replace` equality tests later in this file. The fallback tests can remain separate because their decorators and expected execution differ.

**Review by wjxiz1992 — COMMENTED** [review]
Two P2 test-coverage findings from an exact-head review.

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala (hunk 1 above)
Comment (wjxiz1992) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:267]: A concrete way to cover both production consumers while extending the existing tests is:

```python
# regexp_test.py: use this generator in the existing zero/positive-limit tests
data_gen = mk_str_gen('([bf]o{0,2}:){1,7}') \
    .with_special_case('boo:and:foo') \
    .with_special_case('foo:FOO:FoO')

# append to test_split_re_zero_limit
'split(a, "(?i)foo", 0)',
'split(a, "(?i:foo)", 0)',

# append to test_split_re_positive_limit
'split(a, "(?i)foo", 2)',
'split(a, "(?i:foo)", 5)',
```

```python
# map_test.py: append to test_str_to_map_expr_with_all_regex_delimiters
flag_rows = [
    ('k1xv1PAIRk2xv2',),
    ('k1Xv1pairk2xv2',),
    ('k1Xv1PAIRk2xv2',),
    (None,),
]
assert_gpu_and_cpu_are_equal_collect(
    lambda spark: spark.createDataFrame(flag_rows, ['a']).selectExpr(
        'str_to_map(a, "(?i)pair", "x") as inline_pair',
        'str_to_map(a, "pair", "(?i)x") as inline_key_value',
        'str_to_map(a, "(?i:pair)", "(?i:x)") as scoped_both'),
    conf={'spark.sql.mapKeyDedupPolicy': 'LAST_WIN'})
```

This hits the `limit == 0` and general `n` branches, and independently exercises both the pair and key/value delimiter paths.

File: integration_tests/src/main/python/regexp_test.py (hunk 2 above)
Comment (wjxiz1992) [reply integration_tests/src/main/python/regexp_test.py]: A concrete merge shape is to broaden each existing generator and append the new expressions to that function's current `selectExpr`:

```python
# test_regexp_extract
gen = mk_str_gen('[a-dA-D]{1,3}[0-9]{1,3}/?[a-dA-D]{1,3}')
# append to its existing selectExpr:
'regexp_extract(a, "(?i)([a-c]+)([0-9]+)", 1)',
'regexp_extract(a, "(?i)([a-c]+)([0-9]+)", 2)',
'regexp_extract(a, "([a-c]+)(?i)([a-c]+)", 2)',
'regexp_extract(a, "(?i)(abc)", 1)',

# test_rlike
gen = mk_str_gen('[a-dA-D]{1,4}')
# append to its existing selectExpr:
'a rlike "(?i)abc"',
'a rlike "(?i:abc)"',
'a rlike "(?i)a(?-i)b"',
'a rlike "(?i:(?-i:a)b)"',
# ...the remaining inline/scoped cases from the new block

# test_regexp_replace
gen = mk_str_gen('[a-dA-D]{0,4}')
# append to its existing selectExpr:
'regexp_replace(a, "(?i)abc", "X")',
'regexp_replace(a, "(?i:abc)", "Y")',
'regexp_replace(a, "(?i)a(?-i)b", "Z")',
'regexp_replace(a, "(?i)(abc)", "[$1]")',
```

Then remove `test_regexp_extract_case_insensitive_inline_flag`, `test_rlike_case_insensitive_inline_flag`, and `test_regexp_replace_case_insensitive_inline_flag`. The fallback tests can remain separate.

Comment (igorpeshansky) [reply integration_tests/src/main/python/regexp_test.py]: Thanks for the code examples! Done.

Just to confirm, existing test cases like `test_rlike_null_pattern` use the same setup — should I fold those in as well?

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala (hunk 1 above)
Comment (igorpeshansky) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:267]: Thanks for the code examples! Done.

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala [hunk 3]
```diff
@@ -177,43 +177,79 @@ class RegexParser(pattern: String) {
...
-    RegexGroup(groupType, term)
+    (peek(), groupType) match {
+      case (None, RegexGroup.ScopedFlags(_)) =>
+        throw new RegexUnsupportedException("Unterminated inline flags", Some(pos))
+      case (None, _) => throw new RegexUnsupportedException("Unclosed group", Some(pos))
+      case (Some(')'), RegexGroup.ScopedFlags(flags)) =>
+        consumeExpected(')')
+        RegexInlineFlags(flags)
+      case (Some(ch), _) =>
+        if (ch == ':') consumeExpected(':')
```
Comment (wjxiz1992) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Could you limit the colon consumption to scoped flag groups only? Right now `(:abc)` becomes `(abc)` and `(?::abc)` becomes `(?:abc)`, which changes the regex behavior and can cause CPU/GPU mismatches. Please preserve `:` in normal groups and add regression tests for both cases.

Comment (wjxiz1992) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Reproduced in `spark-shell` on `a7a7a112c5f90195e114602194341bb83e632a46` with `spark.rapids.sql.rLikeRegexRewrite.enabled=false`:

```scala
val query = """
  |SELECT input,
  |       input RLIKE '(:abc)' AS capturing_group,
  |       input RLIKE '(?::abc)' AS noncapturing_group
  |FROM VALUES (':abc'), ('abc') AS t(input)
  |""".stripMargin

val gpu = spark.sql(query)
val gpuRows = gpu.collect()
println(gpu.queryExecution.executedPlan)
gpuRows.foreach(println)

spark.conf.set("spark.rapids.sql.enabled", "false")
val cpu = spark.sql(query)
val cpuRows = cpu.collect()
cpuRows.foreach(println)
```

The GPU plan contains:

```text
GpuProject [... gpurlike (:abc) ..., ... gpurlike (?::abc) ...]
```

Results:

```text
GPU
[:abc,true,true]
[abc,true,true]

CPU
[:abc,true,true]
[abc,false,false]
```

The `abc` row mismatches for both patterns.

Comment (igorpeshansky) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Nice catch, totally missed it before. Done.

> [!NOTE]
> I've only added parser unit tests, not integration tests. Do you feel like this merits end-to-end testing as well, or are the unit tests sufficient?

**Comment by igorpeshansky** [comment]
build

**Review by wjxiz1992 — DISMISSED** [review]

**Comment by wjxiz1992** [comment]
build

**Comment by wjxiz1992** [comment]
I added [databricks] to trigger tests there, just to make sure no side effects.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala (hunk 3 above)
Comment (wjxiz1992) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala]: Unit test is good enough.

File: integration_tests/src/main/python/regexp_test.py (hunk 2 above)
Comment (wjxiz1992) [reply integration_tests/src/main/python/regexp_test.py]: Not necessary in this PR, (rlike_null covers a distinct null/optimizer path, so it should remain separate), but like `test_rlike_embedded_null`, it can be fold in. But we can clean it up when we get a real problem with it. Current PR scope is good.

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by igorpeshansky** [comment]
build

**Comment by wjxiz1992** [comment]
### Databricks Blossom failure investigation

I reproduced this against the exact current PR head, `acc988cbecf80e028d78bd12fab34bf3d5a693e5`, on fresh DBR 14.3 GPU ML and DBR 17.3 GPU ML clusters, using PR JARs built from that head with `jenkins/databricks/build.sh` (`350db143` and `400db173`, respectively).

The latest Blossom build #13996(sent in private) tested a merge commit whose parents are the current base `5a3ed78087d98bd9737eabd92b5fcd8e4a792bb4` and this exact PR head, so this is not a stale-head failure. The only failed top-level stage was `Databricks IT part1`; `test_rlike` failed on both DBR 14.3 and 17.3. The direct error is:

```text
GpuCpuBridgeExpression contains disallowed CPU expressions:
Bridge: GpuCpuBridge(gpuInputs=[a#...], cpuExpression=lower(input[0, string, true]))
CPU Expression Tree Analysis:
  DISALLOWED EXPRESSIONS:
    - Lower (Lower) [NOT ALLOWED]
```

#### Primary cause: a DBR-only optimizer rewrite

On both DBR versions, the newly added expression

```sql
a RLIKE '(?i)abc'
```

is analyzed as `RLIKE`, but a Databricks-specific optimizer rule rewrites it:

```text
RLIKE(a, (?i)abc)
  -- com.databricks.sql.optimizer.SimplifyRLike -->
Contains(lower(a), abc)
```

I verified the rule name in the optimizer batches, invoked the rule directly and obtained the same rewrite, and confirmed that setting

```text
spark.sql.optimizer.excludedRules=com.databricks.sql.optimizer.SimplifyRLike
```

leaves the optimized expression as `RLIKE`.

RAPIDS converts `Contains`, but with the default compatibility configuration it bridges `Lower` to CPU:

```text
GpuProject [... gpucontains(
  GpuCpuBridge(cpuExpression=lower(...)), abc) ...]
```

The strict integration-test plan validation then rejects the unallowlisted `Lower`. This explains the current Blossom failure on both DBR 14.3 and 17.3.

This differs from upstream Apache Spark because `SimplifyRLike` is a Databricks rule. Upstream Spark's `RLike` implementation compiles a Java `Pattern` and calls `matcher(...).find(0)` directly: [Spark 3.5](https://github.com/apache/spark/blob/v3.5.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L445-L486), [Spark 4.0](https://github.com/apache/spark/blob/v4.0.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L488-L529).

#### Secondary DBR 14.3 issue exposed after allowing `Lower`

Enabling incompatible expressions makes `Lower` execute on GPU and clears the primary plan-validation failure on DBR 17.3. It is not sufficient for the whole test on DBR 14.3, however: it exposes CPU/GPU mismatches for position-sensitive bare inline flags.

Examples reproduced on DBR 14.3:

| Input/expression | Java interpreted, whole-stage off, and GPU | DBR 14.3 CPU whole-stage on |
|---|---:|---:|
| `'ABC' RLIKE 'a(?i)b'` | `false` | `true` |
| `'A' RLIKE 'a\|(?i)b'` | `false` | `true` |
| `'AA' RLIKE 'a\|(?i)b'` | `false` | `true` |

DBR's expression codegen stores a precompiled `java.util.regex.Pattern` in a codegen reference object and generates a call to `RuntimeUtils.regexFind(...)`. DBR 14.3 uses Java 8 (`1.8.0_492`). I serialized/deserialized the same patterns directly on that runtime and reproduced the change in behavior: for example, `Pattern.compile("a(?i)b")` does not find a match in `"ABC"` before serialization, but does after deserialization. This matches the fixed OpenJDK issue [JDK-8194667: Regex serialization does not work with match flags](https://bugs.openjdk.org/browse/JDK-8194667), which affects Java 8/9/10 and was fixed in Java 11.

DBR 17.3 uses Java 17 (`17.0.19`). The same serialization probes preserve the inline-flag position there, and whole-stage on/off produce the same results.

Therefore, the DBR 14.3 mismatch is not a GPU-regex semantic defect: GPU agrees with the Java interpreted result. The DBR 14.3 whole-stage CPU reference is changed by the Java 8 serialization bug. It still has to be handled in this parity test because the test compares against the runtime's default CPU execution.

#### Suggested bounded test fix

I would not add `@incompat` to the entire current [`test_rlike`](https://github.com/NVIDIA/cudf-spark/blob/acc988cbecf80e028d78bd12fab34bf3d5a693e5/integration_tests/src/main/python/regexp_test.py#L1046-L1075): that fixes the visible DBR 17.3 failure but exposes the DBR 14.3 mismatch above. Allowing only non-GPU `Lower` also addresses only the first layer.

A safer approach would be:

1. Split the optimizer-rewritten `(?i)abc`-style case from the position-sensitive bare-inline-flag cases.
2. For the optimizer-rewritten case, either:
   - allow DBR's `Lower` if the purpose is to validate the default DBR plan; or
   - exclude `com.databricks.sql.optimizer.SimplifyRLike` in a focused test if the purpose is specifically to exercise `GpuRLike`.
3. For `a(?i)b`, `a|(?i)b`, and similar affected cases on DBR 14.3, either run the focused parity check with `spark.sql.codegen.wholeStage=false`, documenting JDK-8194667, or skip/xfail those cases only on DBR 14.3 while retaining them on Apache Spark and DBR 17.3.
4. Add deterministic special inputs such as `A`, `AA`, and `ABC`; relying only on the random generator can hide the Java 8 behavior.

No source changes were made as part of this investigation.

**Comment by igorpeshansky** [comment]
build

**Review by wjxiz1992 — APPROVED** [review]
